Add docs for MeshGlobalRateLimit policy for Kong Mesh (#5403)
Kong/docs.konghq.com: app/_src/mesh/features/meshglobalratelimit.md

checks if counter for this service is below limits. If it is below limits, it updates the counter and allows the request to pass. If counters are above
limit, deny response is returned. 

{:.note}
> **Note**: You need to remember that configuring global ratelimit will increase yours service response times. Since it needs additional request 
> to ratelimit service and redis.

### Ratelimit service configuration

Beside basic service configuration that is provided on startup, ratelimit service needs limits configuration. Limits configuration
is loaded dynamically from control plane. Ratelimit service uses xDS protocol for this, which is the same protocol that Data Plane Proxy
uses for communicating with Control Plane. Control Plane will periodically compute new limits configuration and send it to ratelimit service.

### Rate limiting algorithm

Rate limit service uses fixed window algorithm. It allocates a new counter for every time unit. Let's assume that we have configured limits to 
10 requests per minute. At the beginning of each minute, ratelimit service will create a new counter. 

<center>
  <img src="/assets/images/docs/mesh/ratelimit-algorithm.png"/>

> </center>

Author: added

When a new request arrives, it updated counters in the window based on the request timestamp. This is simple algorithm but have two main problems about
which you should be aware:

1. counter can be depleted at the begging of the window
2. first counter can be depleted at the end of the window and the second can be depleted at the beginning of the window

The first scenario can be problematic for long time units like hours, and days. Because the limit is reached at the beginning of the time window
you service will not serve any requests for the rest of the time window. To solve this issue, you can split your limit from 60 per hour to 
1 per minute.

The second scenario could result in requests burst around the window switch. In this scenario when you configure limit to 100 requests per minute
you could end up in receiving 200 request in a couple of seconds.

### Multizone deployment

When it comes to multizone deployment, you should deploy ratelimit service in every zone. As for the Redis you have two options:

<center>
  <img src="/assets/images/docs/mesh/ratelimit-service-multizone-multi-redis.png"/>

> </center>

Author: added

First option is to deploy Redis in every zone. In this setup, limits will be applied per zone. Since each zone will have its own counters cache, 
requests will be faster, and it will be easier to distribute your system geographically.

<center>
  <img src="/assets/images/docs/mesh/ratelimit-service-multizone-single-redis.png"/>

> </center>

Author: added

The second option is to deploy single Redis for all your zones. In this setup, the rate limit will be truly global. When deploying single Redis
you need to remember that if your zones are distributed geographically requests to Redis can become slower which could drastically 
increase response times of service you are rate limiting. 

### Ratelimit service security

> #### Securing communication between ratelimit service and Control Plane

Author: changed, this heading will be needed in the future, but we will add this after we implement rest of this

  from:
    - targetRef:
        kind: Mesh
      default:
        local:
          http:
            requestRate:
              num: 10
              interval: 10s
---
apiVersion: kuma.io/v1alpha1
kind: MeshGlobalRateLimit
metadata:
  name: demo-global-rate-limit
  namespace: {{ site.mesh_namespace }}
spec:
  targetRef:
    kind: MeshService
    name: demo-app_kuma-demo_svc_5000
  from:
    - targetRef:
        kind: Mesh
      default:
        http:
          requestRate:
            num: 60
            interval: 1m
```

{% endnavtab %}
{% navtab Universal %}

```yaml
type: MeshRateLimit
mesh: default
name: demo-local-rate-limit
spec:
  targetRef:
    kind: MeshService
    name: demo-app
  from:
    - targetRef:
        kind: Mesh
      default:
        local:
          http:
            requestRate:
              num: 10
              interval: 10s
---
type: MeshGlobalRateLimit
name: demo-global-rate-limit
mesh: default
spec:
  targetRef:
    kind: MeshService
    name: demo-app
  from:
    - targetRef:
        kind: Mesh
      default:
        http:
          requestRate:
            num: 60
            interval: 1m
```

{% endnavtab %}
{% endnavtabs %}

> ### External Service support

Reviewer: ```suggestion
### External service support
```

In order to rate limit requests to [External Service](/mesh/{{page.kong_version}}/policies/external-services) you need to deploy [ZoneEgress](/mesh/{{page.kong_version}}/explore/zoneegress). 

> After deploying Zone Egress, you need to enable mTLS in your mesh and configure zone egress routing. Example mesh config:

Reviewer: ```suggestion
After deploying Zone Egress, you must enable mTLS in your mesh and configure zone egress routing. Here's an example mesh configuration:
```

kind: MeshGlobalRateLimit
metadata:
  name: httpbin-rate-limit
  namespace: {{ site.mesh_namespace }}
spec:
  targetRef:
    kind: MeshService
    name: httpbin
  from:
    - targetRef:
        kind: Mesh
      default:
        http:
          requestRate:
              num: 100
              interval: 1s    
```

{% endnavtab %}
{% navtab Universal %}

```yaml
type: MeshGlobalRateLimit
name: httpbin-rate-limit
mesh: default
spec:
  targetRef:
    kind: MeshService
    name: httpbin
  from:
    - targetRef:
        kind: Mesh
      default:
        http:
          requestRate:
              num: 100
              interval: 1s
```

{% endnavtab %}
{% endnavtabs %}

## Ratelimit Service

{{site.mesh_product_name}} is using **Envoy Rate Limit service reference implementation**. Its source code with documentation can be found [here](https://github.com/envoyproxy/ratelimit).

### Deployment

On Universal, you need to take care of deploying Redis and ratelimit service on your own. On Kubernetes, you could use {{site.mesh_product_name}} Helm charts 
to deploy ratelimit service, but you still need to deploy Redis on your own.

### How to read limits configuration

> When you expose debug port on ratelimit service, you can get access to limits configuration, at `/rlconfig` path.
> Here is an example configuration that you will see something like this:

Reviewer: ```suggestion
When you expose debug port on the ratelimit service, you can get access to the limits configuration at the `/rlconfig` path.
Here is an example configuration:
```

```
default.kuma.io/service_demo-app_kuma-demo_svc_5000: unit=SECOND requests_per_unit=10, shadow_mode: false
```

This config follows a pattern: 
```
mesh.descriptorKey_descriptorValue
```

> Descriptor key will always be `kuma.io/service`, and descriptor value will be your Mesh Service name. 
> Therefore, we can deduct that this config will apply to `demo-app_kuma-demo_svc_5000` service in `default`
> mesh. The second part of the configuration is rather straight forward. We have information of request limit and unit to which 
> this limit will be applied. Shadow mode specifies if request should be limited after reaching the configured limit. If `shadow_mode` is set 
> to `true`, ratelimit service will not deny requests to your service, it will only update ratelimit service metrics.

Reviewer: ```suggestion
The descriptor key will always be `kuma.io/service`, and the descriptor value will be your mesh service name. 
Therefore, we can deduct that this configuration will apply to `demo-app_kuma-demo_svc_5000` service in the `default`
mesh. For the second part of the configuration, we have information about the request limit and unit to which 
this limit will be applied. Shadow mode specifies if request should be limited after reaching the configured limit. If `shadow_mode` is set 
to `true`, the ratelimit service will not deny requests to your service, it will only update the ratelimit service metrics.
```

### Performance improvements

Ratelimit service offers two mechanisms for tweaking its performance. [Local cache for depleted counters](https://github.com/envoyproxy/ratelimit#local-cache) and [Redis query pipelining](https://github.com/envoyproxy/ratelimit#pipelining). 
You should refer to ratelimit service documentation for more concrete documentation and setup.

## Example Ratelimit setup

### Prerequisites

> - Kubernetes cluster on which you will run demo

Reviewer: ```suggestion
- A Kubernetes cluster to run the demo on
```

        port: 6379
      secrets:
        redisAuth:
          Secret: redis
          Key: redis-password
          Env: REDIS_AUTH" > values.yaml
    ```

1. Now we can deploy our control plane

    ```bash
    kumactl install control-plane --values values.yaml | kubectl apply -f -
    ```

1. We now have our control plane with ratelimit setup, so we can install demo app

    ```bash
    kumactl install demo | k apply -f -
    ```

1. After deploying the demo app, you can apply a sample policy:

    ```bash
    echo "apiVersion: kuma.io/v1alpha1
    kind: MeshGlobalRateLimit
    metadata:
      name: demo-rate-limit
      namespace: {{ site.mesh_namespace }}
    spec:
      targetRef:
        kind: MeshService
        name: demo-app_kuma-demo_svc_5000
      from:
        - targetRef:
            kind: Mesh
          default:
            http:
              requestRate:
                num: 10
                interval: 1m
              onRateLimit:
                status: 423
            backend:
              rateLimitService:
                url: http://kong-mesh-ratelimit-service.kong-mesh-system:10003
                timeout: 1s" | kubectl apply -f -
    ```

> 2. We are all set up. Now you can try making few requests to external IP of your gateway, and you will see error after reaching limits. You can find this IP with command:

Reviewer: ```suggestion
2. We are all set up. Now you can try making few requests to the external IP of your gateway and you will see an error after reaching limits. You can find the IP with command:
```

### Multizone deployment

When it comes to multizone deployment, you should deploy the ratelimit service in every zone. As for Redis, you have two options:

<center>
  <img src="/assets/images/docs/mesh/ratelimit-service-multizone-multi-redis.png"/>
</center>
> Figure 3: Diagram of multizone ratelimit setup with Redis per zone.

The first option is to deploy Redis in every zone. In this setup, limits will be applied per zone. Since each zone will have its own counters cache, 
requests will be faster, and it will be easier to distribute your system geographically.

<center>
  <img src="/assets/images/docs/mesh/ratelimit-service-multizone-single-redis.png"/>
</center>
> Figure 4: Diagram of multizone ratelimit setup with global Redis.

The second option is to deploy a single Redis datastore for all your zones. In this setup, the rate limit will be truly global. When deploying a single Redis datastore, remember that if your zones are distributed geographically requests to Redis can become slower which could drastically 
increase response times of service you are rate limiting. 

### Securing communication between ratelimit service and Control Plane

> Communication between the ratelimit service and control plane is encrypted. In addition, zone token authorization is used on Universal and  service account authorization is used on Kubernetes. 

Reviewer 2: I thought this wasn't done yet

Author: It is not done yet, but I've added this part hoping Bart will finish this, I can comment out this section for now and we will add it back when it is done

Author: I commented it out for now

{% navtabs %}
{% navtab Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: Mesh
metadata:
  name: default
spec:
  routing:
    zoneEgress: true
  mtls:
    enabledBackend: ca-1
    backends:
      - name: ca-1
        type: builtin
```

{% endnavtab %}
{% navtab Universal %}

```yaml
type: Mesh
name: default
routing:
  zoneEgress: true
mtls:
  enabledBackend: ca-1
  backends:
    - name: ca-1
      type: builtin
```

{% endnavtab %}
{% endnavtabs %}

After configuring your mesh, you can create your external service:

{% navtabs %}
{% navtab Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: ExternalService
mesh: default
metadata:
  name: httpbin
spec:
  tags:
    kuma.io/service: httpbin
    kuma.io/protocol: http 
  networking:
    address: httpbin.org:443
    tls:
      enabled: true
```

{% endnavtab %}
{% navtab Universal %}

```yaml
type: ExternalService
mesh: default
name: httpbin
tags:
  kuma.io/service: httpbin
  kuma.io/protocol: http
networking:
  address: httpbin.org:443
  tls:
    enabled: true
```

{% endnavtab %}
{% endnavtabs %}

> When applying policies, external services are treated as normal mesh services, so we could configure MeshGlobalRateLimit like this:

Reviewer 2: ```suggestion
When applying policies, external services are treated as normal mesh services, so we configure MeshGlobalRateLimit like this:
```
